---
title: "PCA and Admixture"
author: "Maria Izabel cavassim Alves"
date: '`r Sys.Date()`'
output:
  md_document:
  variant: markdown_github
---

## Population Structure analysis

With the advent of SNP data it is possible to precisely infer the genetic distance across individuals or populations. As written in the book, one way of doing it is by comparing each SNP from each individual against every other individual. This comparison produces the so called: covariance matrix, which in genetic terms means the number of shared polymorphisms across individuals.
There are many ways to visualize this data, in this tutorial you will be exposed to `Principal Component Analysis` and `Admixture` software.

We will use the R package `SNPRelate`, which can easily handle vcf files and do the PCA. If you want to explore a bit more on the functionality of the package access [here](https://www.rdocumentation.org/packages/SNPRelate/versions/1.6.4).

In this first of the exercise please download the vcf file (Allvariants_135_145_chr2.vcf) and the metadata placed in the cluster: `/home/shared/PCA_admixture_data`

```{r setup, echo=F}
knitr::opts_knit$set(root.dir = "/Users/PM/Dropbox/PG2018/exercises/PCA_admixture")
```


```{r, cache = F, message=FALSE}
# Dependencies
#source("http://bioconductor.org/biocLite.R")
#biocLite("SNPRelate")

library(SNPRelate)
library(ggplot2)

# Use setwd to allocate to the directory that you have downloaded your files from cluster.
# I donwloaded my files to: 
setwd("/Users/PM/Dropbox/PG2018/exercises/PCA_admixture")

# Reading the metadata information 
info = read.csv("Sample_meta_data.csv", header = T, sep = ';')

# Setting the directory of the VCF file 
vcf.fn <- "Allvariants_135_145_chr2.vcf"

# Transforming the vcf file to gds format
snpgdsVCF2GDS(vcf.fn, "Allvariants_135_145_chr2_2.gds", method="biallelic.only")

genofile <- snpgdsOpen("Allvariants_135_145_chr2_2.gds",  FALSE, TRUE, TRUE)
pca <- snpgdsPCA(genofile)
summary(pca)
```

**Q.1** How many individuals and snps does this dataset have? What is an eigenvector and an eigenvalue? Hint: Have a look at page 180 of HEG. 

```{r}
eigenvectors = as.data.frame(pca$eigenvect)
colnames(eigenvectors) = as.vector(sprintf("PC%s", seq(1:nrow(pca$eigenvect))))
pca$sample.id = sub("_chr2_piece_dedup", "", pca$sample.id)

# Matching the sample names with their origin and population
eigenvectors$region = info[match(pca$sample.id, info$ENA.RUN),]$region 
eigenvectors$population = info[match(pca$sample.id, info$ENA.RUN),]$population
```

Let's first look at how much of the variance of the data is explained by each eigenvector:
```{r}
# Variance proportion:
pca_percent <- pca$varprop*100

qplot(y = pca_percent, x = seq(1, length(pca$eigenval))) + geom_line() + geom_point() + theme_bw() + xlab("PC's") + ylab("Variance explained (%)") 

```

**Q.2** How many PC's do we need in order to explain 50% of the variance of the data? Can you make an accumulative plot of the variance explained PC?


Now, let's plot the two first PC's and color the datapoints by the origin of each individual sample.
```{r}
ggplot(data = eigenvectors, aes(x = PC1, y = PC2, col = region)) + 
        geom_point(size=3,alpha=0.5) +
        scale_color_manual(values = c("#FF1BB3","#A7FF5B","#99554D")) +
        theme_bw()
```

**Q.2** Try to plot PC2 and PC3. Do you see the same patterns? What is the correlation between PC2 and PC3 (hint use the function cor())?

**Q.3** Try also to color the graph based on population. What do you observe?

Now we will implement LD prunning. 
```{r}
set.seed(1000)

# This function prune the snps with a thrshold of maximum 0.3 of LD
snpset <- snpgdsLDpruning(genofile, ld.threshold=0.3)

# Get all selected snp's ids
snpset.id <- unlist(snpset)

pca_pruned <- snpgdsPCA(genofile, snp.id=snpset.id, num.thread=2)

eigenvectors = as.data.frame(pca_pruned$eigenvect)
colnames(eigenvectors) = as.vector(sprintf("PC%s", seq(1:nrow(pca$eigenvect))))
pca_pruned$sample.id = sub("_chr2_piece_dedup", "", pca$sample.id)

# Matching the sample names with their origin and population
eigenvectors$region = info[match(pca_pruned$sample.id, info$ENA.RUN),]$region 
eigenvectors$population = info[match(pca_pruned$sample.id, info$ENA.RUN),]$population

ggplot(data = eigenvectors, aes(x = PC1, y = PC2, col = region)) + 
        geom_point(size=3,alpha=0.5) +
        scale_color_manual(values = c("#FF1BB3","#A7FF5B","#99554D")) +
        theme_bw() + coord_flip()
```

**Q.4** Implement different LD thresholds (0.1, 0.2, 0.3, 0.4, 0.5). How many SNPs are left after each filtering threshold? Are these SNPs linked? 


Now we are going to convert this GDS file into a plink format, to be later used in the admixture exercise:

```{r}
snpgdsGDS2BED(genofile, "Allvariants_135_145_chr2_pruned.gds", sample.id=NULL, snp.id=snpset.id, snpfirstdim=NULL, verbose=TRUE)
```

Upload the 3 files produced by this last code (**Allvariants_135_145_chr2_pruned.gds.bed**, **Allvariants_135_145_chr2_pruned.gds.bim** and **Allvariants_135_145_chr2_pruned.gds.fam**) to you own folder on the cluster.

# Admixture
Admixture is a program for estimating ancestry in a model based manner from autossomal SNP genotype datasets, where individuals are unrelated. The input format required by the software is in binary PLINK (.bed) file. That is why we converted our vcf file into .bed.

Now with adjusted format and pruned snps, we are ready to run the admixture analysis. We believe that our individuals are derived from three ancestral populations:

```{r, engine = 'bash', eval=F}
../shared/PCA_admixture_data/admixture_linux-1.3.0/admixture Allvariants_135_145_chr2_pruned.gds.bed 3
```

**Q.5** Have a look at the Fst across populations, that is printed in the terminal. Would you guess which populations are Pop0, Pop1 and Pop2 referring to?

After running admixture, 2 outuputs are generated:

- `Q`: the ancestry fractions

- `P`: the allele frequencies of the inferred ancestral populations 

Sometimes we may have no priori about K, one good way of choosing the best K is by doing a cross-validation procedure impletemented in admixture as follow:

```{r, engine = 'bash', eval=F}
for K in 1 2 3 4 5; \
  do ../shared/PCA_admixture_data/admixture_linux-1.3.0/admixture --cv Allvariants_135_145_chr2_pruned.gds.bed $K | tee log${K}.out; done
```

Have a look at the Cross Validation error of each K:

``` {r, engine = 'bash', eval=F}
grep -h CV log*.out
```

Save it in a text file:
``` {r, engine = 'bash', eval=F}
grep -h CV log*.out > CV_logs.txt
``` 

Look at the distribution of CV error. You can download your file to your own computer or run it in the cluster.
```{r}
CV = read.table('CV_logs.txt')
p <- ggplot(data = CV, aes(x = V3, y = V4, group = 1)) + geom_line() + geom_point() + theme_bw() + labs(x = 'Number of clusters', y = 'Cross validation error')

p
#ggsave(p, device = "pdf")
```

**Q.6** What do you understand of Cross validation error? Based on this graph, what is the best K?

Plotting the Q estimates. Choose the K that makes more sense to you.
```{r}
tbl = read.table("Allvariants_135_145_chr2_pruned.gds.3.Q")
ord = tbl[order(tbl$V1,tbl$V2,tbl$V3),]
bp = barplot(t(as.matrix(ord)), 
              space = c(0.2),
              col=rainbow(3),
              xlab="Individual #", 
              ylab="Ancestry",
              border=NA)
```

**Q.7** How many cluster do you identify in this plot? Does that agree with what was found using PCA? 

In the following part of this exercise you will do both analysis (PCA and Admixture) using a different dataset. The data comes from the HAPMAP Consortium, to learn more about the populations studied in this project access [here](http://www.sanger.ac.uk/resources/downloads/human/hapmap3.html). A information file **relationships_w_pops_121708.txt**, as well as **.bim**, **.bed**, **.fam** files are available for the admixture analysis, this dataset is placed in the cluster: **/home/shared/PCA_admixture_data/hapmap**. 

To produce the vcf file out of the these files format you can type:

```{r, engine = 'bash'}
plink --bfile hapmap --recode vcf --out hapmap
```

Answer the same questions as answered in this tutorial and write a report (5 pages maximum) about the results and the analysis you have done. The deadline of the report will be given during the lecture.

```{r, echo = F, eval = F}
# Results of the exercise

CV = read.table('/PM/Downloads/hapmap/CV_logs.txt')
p <- ggplot(data = CV, aes(x = V3, y = V4, group = 1)) + geom_line() + geom_point() + theme_bw() + labs(x = 'Number of clusters', y = 'Cross validation error')
p

tbl = read.table("/Users/PM/Downloads/hapmap/hapmap.4.Q")
ord = tbl[order(tbl$V1,tbl$V2,tbl$V3),]
bp = barplot(t(as.matrix(ord)), 
              space = c(0.2),
              col=rainbow(4),
              xlab="Individual #", 
              ylab="Ancestry",
              border=NA)
# Setting the directory of the VCF file 
vcf.fn <- "/Users/PM/Dropbox/PG2018/exercises/PCA_admixture/hapmap/hapmap.vcf"

# Reading the vcf file
snpgdsVCF2GDS(vcf.fn, "/Users/PM/Dropbox/PG2018/exercises/PCA_admixture/hapmap/data.gds",  method="biallelic.only")

genofile2 <- snpgdsOpen("/Users/PM/Dropbox/PG2018/exercises/PCA_admixture/hapmap/data.gds",  FALSE, TRUE, TRUE)
pca <- snpgdsPCA(genofile2)
summary(pca)

eigenvectors = as.data.frame(pca$eigenvect)

ggplot(data = eigenvectors, aes(x = V1, y = V2)) + 
        geom_point(size=3,alpha=0.5) +
        scale_color_manual(values = c("#FF1BB3","#A7FF5B","#99554D")) +
        theme_bw() + coord_flip()


# Variance proportion:
pca_percent <- pca$varprop*100

qplot(y = pca_percent[1:28], x = seq(1, length(pca$eigenval[1:28]))) + geom_line() + geom_point() + theme_bw() + xlab("PC's") + ylab("Variance explained (%)") 

```


```{r, engine='bash', eval = F, echo = F}
 vcftools --vcf Allvariants_135_145_chr2.vcf --recode --out Allvariants_135_145_chr2_test.vcf --remove-indv ERR1025639_chr2_piece_dedup
```
