{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9b07e7ac-63c3-4f87-995a-8c1c712d4aa3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Building genealogies\n",
    "\n",
    "As seen in the *selective sweep* tutorial, male X chromosomes are easy to use for haplotype-based analysis, as they are haploid and therefore does not require phasing. However, autosomes are diploid, and we would like to use haplotype-based analysis here as well. This can be done through phasing. Multiple phasing methods are known, such as Beagle (already used in the previous tutorial about phasing) and Shapeit. If you are lucky, the dataset you are using has already been phased, which is where we start in this tutorial. We are back to limit us to looking at chr2. The data is taken from 60 individuals in the 1000Genomes project. In this dataset, there are 20 individuals from the following populations: GBR (Brits/Scots), JPT (Japanese), YRI (Yoruban).\n",
    "\n",
    "## Background material\n",
    "\n",
    "We will use the software [Relate](https://myersgroup.github.io/relate/) to infer the genealogies from genome data. It is optimal to read the [software article](https://doi.org/10.1038/s41588-019-0484-x), and for those who want a deeper knowledge of it, the second software paper illustrating updates to the program is [here](https://doi.org/10.1093/molbev/msab174).\n",
    "\n",
    "## How to make this notebook work\n",
    "\n",
    "* In this notebook we will use only `command line Bash`. You need to choose a kernel, that contains a programming language and the necessary packages to run the course material. To choose the right kernel, go on the menu on the top of the page and select `Kernel --> Change Kernel`, and then choose `Bash`.\n",
    "* You can run the code in each cell by clicking on the run cell sign in the toolbar, or simply by pressing <kbd>Shift</kbd>+<kbd>Enter</kbd>. When the code is done running, a small green check sign will appear on the left side.\n",
    "* You need to run the cells in sequential order to execute the analysis. Please do not run a cell until the one above is done running, and do not skip any cells\n",
    "* The code goes along with textual descriptions to help you understand what happens. Please try not to focus on understanding the code itself in too much detail - but rather try to focus on the explanations and on the output of the commands \n",
    "*   You can create new code cells by pressing `+` in the Menu bar above or by pressing <kbd>B</kbd> after you select a cell. \n",
    "\n",
    "\n",
    "## Learning outcomes\n",
    "\n",
    "At the end of this tutorial you will be able to\n",
    "\n",
    "- **Understand** the background theory behind `relate` and **connect** the theory to the application of this tutorial\n",
    "- **Interpret and discuss** the various steps for preparing and analyzing data in `relate`\n",
    "- **Compare and criticize** the usage of multiple vs single populations in the inference of ancestral trees\n",
    "- **Explore, interpret and generalize** simple simulated ancestral trees\n",
    "\n",
    "# Genealogies based on Genome-wide data\n",
    "\n",
    "Knowledge of genome-wide genealogies for many individuals obviously allows for more robust evolutionary analyses for humans and other species - but the computational burden has been unfeasible for long. [Relate](https://myersgroup.github.io/relate/) is a fast and scalable (beyond 10,000 sequences) software that estimates **branch lengths, mutational ages and variable historical population sizes**, as well as **allowing for data errors**. Below is a simple scheme of how the software works - it essentially builds a matrix of distances, where closest sections of chromosomes are more probable to originate from a copying event. From this matrix, a coalescent tree is inferred with dedicated techniques.\n",
    "\n",
    "<img src=\"img/relate_overview.png\" alt=\"arg\" width=\"800\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85cd9eba-2abd-4c70-a35f-6436140599a3",
   "metadata": {},
   "source": [
    "We link to the folder `relate_data` that containes various files needed for the analysis and to the folder containing the necessary software files. We create also the `Results` folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f349be-9b1f-4f40-b6c8-562da8d8152c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln: failed to create symbolic link './relate_data': File exists\n"
     ]
    }
   ],
   "source": [
    "ln -s ../../Data/relate_data\n",
    "ln -s ../../Data/software/relate_v1.1.9_x86_64_static relate\n",
    "mkdir -p Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b81fb989-3273-49af-8259-5b23f69079ef",
   "metadata": {},
   "source": [
    "For the analysis we use the following files:\n",
    "\n",
    "- `20140520.chr2.strict_mask.fasta` is a mask containing areas that either have abnormal read depth or has been identified to contain repetitive elements.\n",
    "- `genetic_map_chr2_combined_b37.txt` is a recombination map.\n",
    "- `human_ancestor_2.fa` is the ancestral state of every site, based on an alignment with Gorilla, Chimp and Human genomes.\n",
    "- `60_inds.txt` is a metadata file detailing the population and region for each sample.\n",
    "- `chr2_130_145_phased.vcf.gz` is the phased genotype vcf.\n",
    "\n",
    "Relate does not accept the standard vcf file format, but instead uses a haps/sample format. You can read up on in the Relate documentation. The authors have been so kind as to supply a script to transform it.\n",
    "\n",
    "**Note that input files in Relate always have to be supplied without file extensions.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1254a0b3-3194-4d03-83f6-f2d19d9be803",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20140520.chr2.strict_mask.fasta\n",
      "50_GBR_inds.txt\n",
      "50_GBR_sample_list.txt\n",
      "60_inds.txt\n",
      "60_sample_list.txt\n",
      "ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\n",
      "ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi\n",
      "ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.vcfidx\n",
      "chr2_130_145_phased.vcf.gz\n",
      "chr2_GBR_phased.vcf.gz\n",
      "genetic_map_chr2_combined_b37.txt\n",
      "human_ancestor_2.fa\n"
     ]
    }
   ],
   "source": [
    "ls relate_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cee5d23a-f2df-46ce-8f90-3c851c152564",
   "metadata": {},
   "source": [
    "## Input preparation\n",
    "\n",
    "The vcf is here converted to another file format (haplotype file format) as just mentioned. If you want to know how it is structured, you can read about it [here](https://www.cog-genomics.org/plink/2.0/formats#haps)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cc64149-9d9c-4adb-9f4e-b91456fa5285",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------------------------------------\n",
      "Converting vcf file format to haps/sample file format..\n",
      "Output written to Results/chr2.haps and Results/chr2.sample.\n",
      "CPU Time spent: 1.036033s; Max Memory usage: 2.508Mb.\n",
      "---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "relate/bin/RelateFileFormats --mode ConvertFromVcf --haps Results/chr2.haps --sample Results/chr2.sample -i relate_data/chr2_130_145_phased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7b4c5d-6db4-496f-a63a-4c73767cbcbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************************\n",
      "Parameters passed to script:\n",
      "haps        = Results/chr2.haps\n",
      "sample      = Results/chr2.sample\n",
      "ancestor    = relate_data/human_ancestor_2.fa\n",
      "output      = Results/prep.chr2\n",
      "mask        = relate_data/20140520.chr2.strict_mask.fasta\n",
      "********************************\n",
      "---------------------------------------------------------\n",
      "Removing non-biallelic SNPs..\n",
      "Removed 0 non-biallelic SNPs.\n",
      "Output written to Results/prep.chr2_biall.haps.\n",
      "CPU Time spent: 0.060300s; Max Memory usage: 0.92Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Detemining ancestral allele and flipping SNPs if necessary... \n",
      "Had to remove 10485 SNPs because of non-matching nucleotides\n",
      "Number of flipped SNPs is 12662.\n",
      "Output written to Results/prep.chr2_ancest.haps.\n",
      "CPU Time spent: 1.447338s; Max Memory usage: 410.672Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "relate/scripts/PrepareInputFiles/PrepareInputFiles.sh: line 120: file: command not found\n",
      "---------------------------------------------------------\n",
      "Filtering haps using mask.. \n",
      "Removed 19650 SNPs.\n",
      "Output written to Results/prep.chr2_filtered.haps and Results/prep.chr2_filtered.dist\n",
      "CPU Time spent: 1.484666s; Max Memory usage: 412.308Mb.\n",
      "---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "relate/scripts/PrepareInputFiles/PrepareInputFiles.sh --haps Results/chr2.haps --sample Results/chr2.sample \\\n",
    "                                                      --ancestor relate_data/human_ancestor_2.fa \\\n",
    "                                                      --mask relate_data/20140520.chr2.strict_mask.fasta \\\n",
    "                                                      -o Results/prep.chr2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12dce69a-6afa-4475-8ed3-3b18c5a835ea",
   "metadata": {},
   "source": [
    "Look at the outputs above, and answer to the following questions.\n",
    "- **Q1: How many SNPs were removed due to non-matching nucleotides?**\n",
    "- **Q2: How many were removed due to the mask?**\n",
    "\n",
    "## Estimation of ancestral trees\n",
    "\n",
    "Now, the input is fully prepared, and the inference of genealogies with `Relate` can run. This should take less than a minute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32509f74-af58-47cc-8c90-930c3f4590a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "*********************************************************\n",
      "---------------------------------------------------------\n",
      "Relate\n",
      " * Authors: Leo Speidel, Marie Forest, Sinan Shi, Simon Myers.\n",
      " * Doc:     https://myersgroup.github.io/relate\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Using:\n",
      "  Results/prep.chr2.haps.gz\n",
      "  Results/prep.chr2.sample.gz\n",
      "  relate_data/genetic_map_chr2_combined_b37.txt\n",
      "with mu = 1.25e-08 and 2Ne = 30000.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Parsing data..\n",
      "Warning: Will use min 0.00035GB of hard disc.\n",
      "CPU Time spent: 0.682457s; Max Memory usage: 23Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Read 120 haplotypes with 67935 SNPs per haplotype.\n",
      "Expected minimum memory usage: 0.61Gb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Starting chunk 0 of 0.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Painting sequences...\n",
      "CPU Time spent: 1.655027s; Max Memory usage: 23Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Estimating topologies of AncesTrees in sections 0-0...\n",
      "CPU Time spent: 5.264306s; Max Memory usage: 8.2e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Propagating mutations across AncesTrees...\n",
      "CPU Time spent: 5.952430s; Max Memory usage: 8.2e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Inferring branch lengths of AncesTrees in sections 0-0...\n",
      "CPU Time spent: 46.492151s; Max Memory usage: 8.2e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Combining AncesTrees in Sections...\n",
      "CPU Time spent: 46.787977s; Max Memory usage: 8.2e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Finalizing...\n",
      "Number of not mapping SNPs: 529\n",
      "Number of flipped SNPs    : 274\n",
      "CPU Time spent: 47.764615s; Max Memory usage: 8.2e+02Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Done.\n",
      "---------------------------------------------------------\n",
      "*********************************************************\n",
      "\n"
     ]
    }
   ],
   "source": [
    "relate/bin/Relate --mode All -m 1.25e-8 -N 30000 --haps Results/prep.chr2.haps.gz --sample Results/prep.chr2.sample.gz \\\n",
    "                  --map relate_data/genetic_map_chr2_combined_b37.txt \\\n",
    "                  -o chr2_relate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fad5cb5-a500-44c5-a59a-cc70f84497c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv chr2_relate.* ./Results/ #moving output into Results folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337d3c05-294f-4675-bf8f-103d68f2b672",
   "metadata": {},
   "source": [
    "- **Q3: How many SNPs are left per haplotype?**\n",
    "\n",
    "## Estimation of population size\n",
    "\n",
    "The lengthiest process is this step, in which population size is estimated, and the population size is re-estimates branch lengths. This takes up to 20 minutes. While you wait, open the notebook `07b_interactive_exercise` and try to answer to the questions trying out the interactive interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f36978-92fb-45f0-81af-e2e6307a6ffc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************************\n",
      "Parameters passed to script:\n",
      "input         = Results/chr2_relate\n",
      "poplabels     = relate_data/60_inds.txt\n",
      "mu            = 1.25e-8\n",
      "years_per_gen = 28\n",
      "output        = Results/popsize\n",
      "num_iter      = 10\n",
      "threshold     = 0\n",
      "Maximum number of threads: 1\n",
      "********************************\n",
      "---------------------------------------------------------\n",
      "Extracting dist file from Results/chr2_relate.mut ... \n",
      "CPU Time spent: 0.083929s; Max Memory usage: 58.744Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Removing trees with few mutations from Results/chr2_relate.mut ... \n",
      "CPU Time spent: 2.502428s; Max Memory usage: 282.068Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.765854s; Max Memory usage: 59.904Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.845644s; Max Memory usage: 168.776Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.749195s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 68.056845s; Max Memory usage: 168.892Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.745532s; Max Memory usage: 59.904Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 66.573337s; Max Memory usage: 168.888Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.756833s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.010154s; Max Memory usage: 168.892Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.750237s; Max Memory usage: 59.904Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.448596s; Max Memory usage: 168.888Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.740143s; Max Memory usage: 59.904Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.362031s; Max Memory usage: 168.892Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.741483s; Max Memory usage: 59.904Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 66.451561s; Max Memory usage: 168.892Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.750135s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.161951s; Max Memory usage: 168.892Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.789958s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.409804s; Max Memory usage: 168.724Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.815423s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Sampling branch lengths for Results/popsize ...\n",
      "[100%]\n",
      "CPU Time spent: 67.701972s; Max Memory usage: 168.656Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rates for Results/popsize...\n",
      "[100%]\n",
      "CPU Time spent: 0.771737s; Max Memory usage: 59.972Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating coalescence rate for Results/popsize ...\n",
      "CPU Time spent: 2.466906s; Max Memory usage: 119.632Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Finalizing coalescence rate...\n",
      "        |                                                          \n",
      "6.90e+04|**                                                        \n",
      "        |                                                          \n",
      "        |  **                                                      \n",
      "        |                              ****                **      \n",
      "        |    **                            **            **        \n",
      "        |      ********              **      **  ****  **          \n",
      "        |                          **          **    **            \n",
      "        |              **        **                                \n",
      "        |                **  **                                    \n",
      "        |                  **  **                                  \n",
      "1.51e+03|                                                    **    \n",
      "        |                                                          \n",
      "        -----------------------------------------------------------\n",
      "        3.57e+01                                            3.57e+05\n",
      "                                 generations ago\n",
      "CPU Time spent: 2.488972s; Max Memory usage: 119.632Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating average mutation rate for Results/popsize ...\n",
      "        |                                                          \n",
      "1.81e-08|  **                                                      \n",
      "        |**  ******                                                \n",
      "        |          ****                                            \n",
      "        |              **                                          \n",
      "        |                **                                        \n",
      "        |                  ****                                    \n",
      "        |                      ****            **                  \n",
      "        |                          ************  **                \n",
      "        |                                                          \n",
      "        |                                          ****            \n",
      "1.27e-08|                                              ********    \n",
      "        |                                                          \n",
      "        -----------------------------------------------------------\n",
      "        3.57e+01                                            3.57e+05\n",
      "                                 generations ago\n",
      "CPU Time spent: 0.846613s; Max Memory usage: 118.66Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Reinferring branch lengths for Results/chr2_relate ...\n",
      "[100%]\n",
      "CPU Time spent: 62.879230s; Max Memory usage: 110.744Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Finalizing coalescence rate...\n",
      "CPU Time spent: 0.006047s; Max Memory usage: 2.616Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "Transformation introduced infinite values in continuous x-axis \n",
      "\u001b[?25hnull device \n",
      "          1 \n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "relate/scripts/EstimatePopulationSize/EstimatePopulationSize.sh -i Results/chr2_relate \\\n",
    "                                                                -m 1.25e-8 --poplabels relate_data/60_inds.txt \\\n",
    "                                                                -o Results/popsize --threshold 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3944ba8-438f-4f1d-a6de-dd9078bcfd6e",
   "metadata": {},
   "source": [
    "- **Q4: At the end, Relate outputs estimated mutation rate and coalescence times along the region - can this tell us anything?**\n",
    "\n",
    "## Selection\n",
    "\n",
    "With the coalescence rates estimated, it is possible to detect selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "306833d8-fac0-4566-b195-afca335171db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************************\n",
      "Parameters passed to script:\n",
      "input         = Results/popsize\n",
      "mu            = 1.25e-8\n",
      "years_per_gen = 28\n",
      "output        = Results/selection_relate\n",
      "********************************\n",
      "---------------------------------------------------------\n",
      "Calculating frequency through time for Results/popsize.\n",
      "CPU Time spent: 1.368211s; Max Memory usage: 117.668Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "---------------------------------------------------------\n",
      "Calculating evidence of selection for Results/selection_relate.\n",
      "CPU Time spent: 2.730782s; Max Memory usage: 1.16Mb.\n",
      "---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "relate/scripts/DetectSelection/DetectSelection.sh -i Results/popsize -m 1.25e-8 \\\n",
    "                                                  --poplabels relate_data/60_inds.txt \\\n",
    "                                                  -o Results/selection_relate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d33933c3-9557-4933-929d-4feb61d32c2d",
   "metadata": {},
   "source": [
    "At this point, we are detecting selection based on three distinct populations. \n",
    "\n",
    "**Q5: Consider whether the underlying assumptions for this make sense.**\n",
    "\n",
    "To find the sites with the strongest selection, we will sort based on col 35 (which is the one with the log10 p-value for selection). The first command is used for sorting. You will see the positions in the first column of the output.\n",
    "\n",
    "In the second command we sort again, but isolate the first column (genomic position) and column 35 (log10 p-value for selection), and then keep only the first 10 rows (excluding the names `pos` and `when_mutation_has_freq2` in the first row). Change the number `11` in the command `sed -n '2,11'p` to choose how many lines to print."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f5106ae-6eb5-4173-834a-80dab8462019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pos rs_id 3571428.500000 357142.937500 257030.656250 184981.281250 133128.375000 95810.585938 68953.507812 49624.847656 35714.292969 25703.064453 18498.126953 13312.836914 9581.057617 6895.350098 4962.484863 3571.429199 2570.306396 1849.812622 1331.283691 958.105774 689.534973 496.248444 357.142914 257.030640 184.981262 133.128357 95.810577 68.953499 49.624844 35.714287 0.000000 when_DAF_is_half when_mutation_has_freq2\n",
      "143026077 rs59280189 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -6.09684 -5.36534 -4.28703 -3.68305 -3.25563 -2.23481 -1.05641 -1.09153 -0.66056 -0.448427 -0.309981 -0.0787622 0 0 0 0 0 -1.80559 -6.09684\n",
      "136608646 rs4988235 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -5.01126 -5.96369 -5.76504 -5.91704 -4.49533 -1.23787 0 0 0 0 -6.35246 -5.47079\n",
      "136616754 rs182549 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -5.01126 -5.96369 -5.76504 -5.91704 -4.49533 -1.23787 0 0 0 0 -6.35246 -5.47079\n",
      "131110097 rs6750645 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -4.12253 -3.61525 -2.12713 -1.15708 -1.03286 -0.692444 -1.17085 -1.04607 -0.773465 -0.381002 -0.0568152 0 0 0 -1.28724 -5.46694\n",
      "141653595 rs12469352 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -4.04941 -1.95932 -1.53461 -1.60475 -2.50727 -1.37845 -0.704867 -0.484479 0 0 0 0 -1.83952 -5.23884\n",
      "130500081 rs13017715 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -3.935 -2.65248 -1.38281 -1.48468 -0.936435 -0.526179 -0.558078 -0.265231 -0.0159609 0 0 0 0 0 -2.22373 -5.23617\n",
      "136169799 rs6742013 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -5.82528 -6.7453 -5.40901 -2.27544 -1.81653 -2.41799 -1.72666 -1.78017 -0.887542 -0.218208 0 0 0 -1.91098 -5.21295\n",
      "130498804 rs10928929 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -4.5755 -5.78962 -4.54734 -3.94768 -3.19114 -3.34317 -1.45072 -0.908191 -0.626947 -0.182633 -0.0296485 -0.0107089 -0.0107089 0 0 0 -2.99375 -5.18567\n",
      "130499359 rs4233597 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 -4.5755 -5.78962 -4.54734 -3.94768 -3.19114 -3.34317 -1.45072 -0.908191 -0.626947 -0.182633 -0.0296485 -0.0107089 -0.0107089 0 0 0 -2.99375 -5.18567\n",
      "sort: write failed: 'standard output': Broken pipe\n",
      "sort: write error\n"
     ]
    }
   ],
   "source": [
    "sort -k35r Results/selection_relate.sele | head -n 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f137ae7-3f42-4815-a3f0-2e98683125f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "143026077 -6.09684\n",
      "136608646 -5.47079\n",
      "136616754 -5.47079\n",
      "131110097 -5.46694\n",
      "141653595 -5.23884\n",
      "130500081 -5.23617\n",
      "136169799 -5.21295\n",
      "130498804 -5.18567\n",
      "130499359 -5.18567\n",
      "139393537 -5.03088\n"
     ]
    }
   ],
   "source": [
    "sort -k35r Results/selection_relate.sele | cut -f1,35 -d' ' | sed -n '2,11'p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66f1dcf2-ee13-474d-94e6-9ff0df0ffaea",
   "metadata": {},
   "source": [
    "Now we know position and p-value for the highest positive selections. Try plotting some of the sites with the following command. Use `-o` to determine the output name of the plot, and `bp_of_interest` for the genomic position. Below is an example for position `143026077`. We change the output name according to position, so that we will not overwrite the plots. The output is a pdf plot with name `Results/tree_POSITION.pdf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "745cdab7-3cbb-47e3-ad1c-559c6086cc6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************************\n",
      "Parameters passed to script:\n",
      "haps           = Results/chr2.haps\n",
      "sample         = Results/chr2.sample\n",
      "poplabels      = relate_data/60_inds.txt\n",
      "anc            = Results/popsize.anc\n",
      "mut            = Results/popsize.mut\n",
      "years_per_gen  = 28\n",
      "bp_of_interest = 143026077\n",
      "output         = Results/tree_143026077.pdf\n",
      "********************************\n",
      "Loading required package: dplyr\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\u001b[?25hLoading required package: ggplot2\n",
      "\u001b[?25hLoading required package: cowplot\n",
      "Installing package into ‘/usr/local/spark-3.2.1-bin-hadoop3.2/R/lib’\n",
      "(as ‘lib’ is unspecified)\n",
      "trying URL 'http://cran.us.r-project.org/src/contrib/cowplot_1.1.1.tar.gz'\n",
      "Content type 'application/x-gzip' length 1353271 bytes (1.3 MB)\n",
      "==================================================\n",
      "downloaded 1.3 MB\n",
      "\n",
      "* installing *source* package ‘cowplot’ ...\n",
      "** package ‘cowplot’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (cowplot)\n",
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpSu166h/downloaded_packages’\n",
      "Loading required package: cowplot\n",
      "Warning message:\n",
      "In library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE,  :\n",
      "  there is no package called ‘cowplot’\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h---------------------------------------------------------\n",
      "Get tree at BP 143026077...\n",
      "CPU Time spent: 0.686335s; Max Memory usage: 117.704Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "\u001b[?25h---------------------------------------------------------\n",
      "Get list of mutations that map to tree at BP 143026077...\n",
      "CPU Time spent: 1.072413s; Max Memory usage: 206.456Mb.\n",
      "---------------------------------------------------------\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead. \n",
      "\u001b[?25h\u001b[?25h\u001b[?25hnull device \n",
      "          1 \n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "POSITION=143026077\n",
    "relate/scripts/TreeView/TreeView.sh --haps Results/chr2.haps --sample Results/chr2.sample \\\n",
    "                                    --anc Results/popsize.anc --mut Results/popsize.mut \\\n",
    "                                    --poplabels relate_data/60_inds.txt --years_per_gen 28 \\\n",
    "                                    -o Results/tree_$POSITION --bp_of_interest $POSITION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "387ba4b0-1187-42e1-a5de-785ed58cbfb1",
   "metadata": {
    "tags": []
   },
   "source": [
    "- **Q6: Is it the same populations in which the sweep seem to occur?**\n",
    "\n",
    "- **Q7: The various populations are generally not fully split. Consider why Relate can end up inferring this.**\n",
    "\n",
    "- **Q8: For a site which seems promising, figure out what the nearby genes code for.**\n",
    "\n",
    "## Exercise: perform the analysis for a single population\n",
    "\n",
    "In general, selection scans of this kind will perform better if it is for a single population. Try rerunning the analysis with pure brits. Copy and paste code from the previous commands and adapt them to the new data\n",
    "\n",
    "Necessary files for this are as well in the `relate_data` folder, and are \n",
    "\n",
    "- relate_data/50_GBR_inds.txt\n",
    "- relate_data/chr2_GBR_phased.vcf.gz \n",
    "\n",
    "which substitute the files `60_inds.txt` and `chr2_130_145_phased.vcf.gz`. The rest of the files needed are still the same as before.\n",
    "\n",
    "When you are done, answer to the following questions:\n",
    "- **Q9: Compared to the results in Q6, do you get different or similar results?**\n",
    "- **Q10: Try plotting some of the same sites, and see whether the trees (when considering GBR) are similar.**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
